### DISCUSSION TITLE
Order execution for builtin ops without input.

### DISCUSSION THREAD

**Discussion by 3cham** [dagster-io/dagster#7926] [Q&A]
Hi, I try to get my first pipeline in dagster working and have following question. I have asked in the slack channel but it might be missed: 

I want to have two built-in ops executed in order, not by data dependency. The way I have found in the doc is to extend the ins parameter with In(Nothing). However, these two built-in op (let's say create_shell_command_op) don't have the ins input. I think of wrapping these two ops inside an op factory, but I could not invoke them in the inner op definition. The only way is to copy the original implementation to a new op definition and extend it with another input. And that does not feel right. 

What have I missed to get this fundamental scenario working? Thanks in advance.

**Comment by HusseinKabbout** [comment]
I would also be interested to know how the execution order in a job is defined in dagster. In the documentation, I could only find [this](https://docs.dagster.io/concepts/ops-jobs-graphs/jobs-graphs#order-based-dependencies-nothing-dependencies). Is using the `In(Nothing)` input really the only way of ordering ops in a job?

**Comment by jamiedemaria** [comment]
Hey there. In general dependencies between ops are defined by the data passed between the ops. This allows us to run ops in parallel if they don't have a dependency on each other. However, sometimes you run into a case when you need `op_A` to run before `op_B` but `op_A` doesn't actually return any data. This is where the Nothing type comes in. it allows you to return `Nothing` from `op_A` and pass it to `op_B` to ensure that `op_B` runs after `op_A`. 

For your specific use case, can you tell us what built in ops you are trying to use and the order they need to run in?

**Reply by 3cham** [reply]
Hey @jamiedemaria, in my case, I need to run a dataproc job in gcp. It is quite straightforward, however for the job config I need the main python file being uploaded into a gcs bucket. That is done in another op.

The problem is, the dataproc_op and many other built-in ops don't consume any input. So what should I do in this case to have the order of execution guaranteed? Many thanks

**Reply by jamiedemaria** [reply]
One thing you can try is wrapping each built-in op in a graph and then passing your ordering dependencies to the graphs within the larger job

**Reply by 3cham** [reply]
that sounds like much work to be done for achieving just a simple pipeline logic. Is it possible to extend a default input (e.g. after=) for every op? Is there any consideration about it already?

**Reply by gibsondan** [reply]
Hi @3cham - it sounds like this feature request is relevant: https://github.com/dagster-io/dagster/issues/4274

The example here is the way to model order-based dependencies in dagster currently: https://docs.dagster.io/concepts/ops-jobs-graphs/jobs-graphs#order-based-dependencies-nothing-dependencies

**Reply by jayhale** [reply]
@gibsondan this docs link is dead now - is this documented elsewhere?

**Reply by yashafromrussia** [reply]
This is documented here https://docs.dagster.io/concepts/ops-jobs-graphs/graphs#defining-nothing-dependencies
